Start the release of the `ocaml` command for other purposes

dra27 opened the pull request:
This was discussed at a core devs' meeting before 4.14 was branched, but I didn't write the commit until too close to the branching date.

Two (semi-distant) future possibilities:
- It might be nice if the two toplevels (`ocaml` and `ocamlnat`) agreed on a name and were `name.byte` and `name.opt`
- It might also be nice if `ocaml` could have sub-commands, or other meanings than invoking the bytecode interpreter (especially as these days a significant portion of the community actually consume the toplevel via `dune utop`). `cargo`, `drom`, etc.

Both of those are difficult and bike-sheddy discussions for another day, but to be even possible, the latter requires a small tweak to the existing bytecode toplevel which is what this PR addresses. We need `ocaml command` to stop meaning "run the file called `command` from the current directory".

This PR - for consideration for 5.0 - deprecates that syntax (it presently also includes a minor refactor commit from #11198). If you have a script called `hello`, with this PR it will now display:
```
dra@thor:~/ocaml$ PATH=~/ocaml/install/bin:$PATH ocaml hello
File "_none_", line 1:
Alert ocaml_deprecated_cli: Running ocaml where the first argument
is an implicit basename with no extension (e.g. ocaml script-file) is
deprecated. Either rename the script (ocaml script-file.ml) or qualify
the basename (ocaml ./script-file)
Hello, world
```
This alert may be silenced in three ways:
1. Rename the file to include an extension (e.g. `hello.ml`)
2. Make the filename explicit (i.e. `./hello`)
3. Use the `-` argument to stop it from being the first parameter (i.e. `ocaml - hello`)

Note in particular that the third form provides an escape hatch if there are any Unix systems still out there which don't pass absolute path to the script when used in a shebang, e.g. `hello`:

```
#!/usr/bin/ocaml
print_endline "Hello, world"
```
On modern Unix systems, the path to the script is passed either as a full path (e.g. `/home/dra/ocaml/hello`) or at least as a `/dev/fd/N` but the shebang could be changed to `#!/usr/bin/ocaml -` for a system which chooses to pass the script name verbatim (i.e. if you run `hello`, and `hello` is resolved and in the current directory, there have been Unix systems which passed `hello` instead of `./hello` as the script filename; most now prepend the `./` or give the absolute path.

dra27 committed:
Deprecate implicit basenames with no extension

In order to free up the ocaml command to be able to be used for other
things, require that a script argument to ocaml (i.e. an anonymous
argument) is either a path (e.g. ./script) or has an extension
(e.g. ocaml script.ml).

This only applies to the first argument.

dbuenzli commented:
> 3\. Use the `-` argument to stop it from being the first parameter (i.e. `ocaml - hello`)

I don't think that's a great idea. Using `-` to mean standard input or output is quite an established convention in cli tools. So that would be rather confusing.

I think you should rather use `--` which is the convention to mean interpret what follows as positional arguments; here the positional arguments to give to `ocaml [CMD]` when `CMD` is unspecified.

dra27 commented:
I'm sorry to say that it's not what I'm proposing, it's what's already there 🙈

```
dra@thor:~$ ocaml -help | grep Treat
  - <file>  Treat <file> as a file name (even if it starts with `-')
```

(although I did manage to add the `--` form to `ocamlrun` in 4.13)

dbuenzli commented:
Ah sorry :-)

avsm commented:
I'm broadly in favour of this, but just need to check hashbang behaviour on OpenBSD to check if it passes a fully qualified path or not.

dra27 commented:
> I'm broadly in favour of this, but just need to check hashbang behaviour on OpenBSD to check if it passes a fully qualified path or not.

@shindere had a quick check on one of the Jenkins boxes (OpenBSD 6.7), and it appears to behave as Linux: if `foo` is a shebang script in the current directory, then the script argument passed to the interpreter is `./foo` (even if invoked as `PATH=.$:PATH foo`).

dra27 commented:
Rebased with an update to the formatting of the message for line breaks.

Octachron commented on toplevel/byte/topmain.ml:
```diff
@@ -203,8 +203,15 @@ end)
+      None
+    else
+      Warnings.deprecated_script_alert !argv.(1)
+  in
```
Should we also add this alert to `ocamlnat` in order to reduce code divergence between the two toplevels?

dra27 replied:
I think this divergence is ok (especially as we still don't _install_ `ocamlnat` by default), just because if a user wants to run `ocamlnat foo` then we don't need to care (because it's not `ocaml foo`). _If_ we decide on a new name for invoking just the toplevel (let's say `ocamlrepl`, just to have a name) then I would propose that `ocamlrepl foo` would be OK as well, because `ocamlrepl` has clearly said what's supposed to happen - i.e. it's just `ocaml foo` which wants the alert.

I have no objection to ocamlnat having the alert, obviously, that's just the reasoning behind why I hadn't done it in the "first cut"!

shindere commented:
I understand your points, @Êra27, but I still tend to agree with
@Octachron and would prefer more similar behavours of `ocaml` and
`ocamlnat`, for a functional reason (both of them are toplevels
currently so it seems to make sense that their behaviours are as similar
as possible) and for the sake of sharking code where possible.

That's just an opnion, I am fine if it's not the way things get finally
done.

dra27 committed:
Have ocamlnat display the warning too

dra27 commented:
Rebased to remove the commit already merged in #11198. Refactored slightly, as there was logic embedded in the alert in `Warnings` - it makes more sense have the logic in `Topcommon` and the alert displaying itself in `Location`, along with the other alert displaying logic.

Finally, the third commit extends the alert to ocamlnat.

shindere commented:
In terms of code this PR looks good to me.

The Changes entry seems to miss a ` character after ocalmnat script.

Re: the name of the arguemnt to `deprecated_script_alert`, how about using
`program` or so rather than `command` which looks kind of ambiguous to me
since it can also be understood as referring ot the argument passed to the
program.

dra27 commented:
Thanks, @shindere! Rebased and commits squashed - assuming you're happy, @Octachron, I'll merge with CI.

Octachron reviewed (approved):
The current code looks good to me. At some point, it might useful to split the `Location` module in more meaningful components, but that will be for later.

shindere commented:
Florian Angeletti (2022/06/07 01:16 -0700):
> @Octachron approved this pull request.
>
> The current code looks good to me. At some point, it might useful to
> split the `Location` module in more meaningful components, but that
> will be for later.

Do we have a way to track this kind of improvemnets one may want to do
one day?

Would it be appropriate to track them with issues and a label?
